src/toxcore/dht/ping_checker.rs

```diff
@@ -0,0 +1,221 @@
+*/
+
+
+/*!
+Manage ping_id.
+Generate ping_id on request packet, check ping_id on response packet.
+*/
+
+use std::io::{ErrorKind, Error};
+use std::collections::HashMap;
+use std::time::{Duration, Instant};
+
+use toxcore::crypto_core::*;
+use toxcore::dht::packed_node::*;
+
+/// Ping timeout in seconds
+pub const PING_TIMEOUT: u64 = 5;
+
+/// Number of DHT ping entries.
+pub const DHT_PING_ENTRY_SIZE: u64 = 512;
+
+/// Entry type of PingChecker
+#[derive(Clone, Debug, PartialEq)]
+struct PingCheckerEntry {
+    data: Vec<PackedNode>,
+    time: Instant,
+    ping_id: u64,
+}
+
+impl PingCheckerEntry {
+    /// Check if it is timedout
+    pub fn is_timeout(&self, timeout: Duration) -> bool {
+        self.time.elapsed() > timeout
+    }
+}
+
+/// Generate ping_id, check if ping_id is correct or timedout
+#[derive(Clone, Debug, PartialEq)]
+pub struct PingChecker {
+    entries: HashMap<u64, PingCheckerEntry>,
+    last_added: u64,
+    last_deleted: u64,
+    total_size: u64,
+}
+
+impl PingChecker {
+    /// Create new PingChecker object
+    pub fn new(total_size: u64) -> PingChecker {
+        assert!(total_size > 0);
+
+        PingChecker {
+            entries: HashMap::new(),
+            last_added: 0,
+            last_deleted: 0,
+            total_size,
+        }
+    }
+
+    // Clear timed out entries
+    fn clear_timedout(&mut self, timeout: Duration) {
+        while self.last_deleted != self.last_added {
+            if let Some(ping_entry) = self.entries.get(&self.last_deleted) {
+                if !ping_entry.is_timeout(timeout) {
+                    break;
+                }
+            }
+
+            self.entries.remove(&self.last_deleted);
+            self.last_deleted = (self.last_deleted + 1) % self.total_size;
+        }
+    }
+
+    /// Add a data to the PingChecker and return a new ping_id.
+    pub fn add(&mut self, data: Vec<PackedNode>) -> u64 {
+        let timeout_dur = Duration::from_secs(PING_TIMEOUT);
+        self.clear_timedout(timeout_dur);
+
+        let mut ping_id = random_u64();
+        ping_id = ping_id - (ping_id % self.total_size) + self.last_added;
+
+        if ping_id == 0 {
+            ping_id = self.total_size;
+        }
+
+        let new_ping_checker_entry = PingCheckerEntry {
+            data,
+            time: Instant::now(),
+            ping_id,
+        };
+
+        self.entries.insert(self.last_added, new_ping_checker_entry);
+
+        self.last_added = (self.last_added + 1) % self.total_size;
+
+        ping_id
+    }
+
+    /// Check if ping_id is valid and not timed out.
+    /// On success, return the data stored.
+    pub fn is_valid(&mut self, ping_id: u64, timeout: Duration) -> Result<Vec<PackedNode>, Error> {
+        if ping_id == 0 {
+            return Err(Error::new(ErrorKind::Other, "Given ping_id is 0"))
+        }
+
+        let index = ping_id % self.total_size;
+
+        let ping_checker_entry = match self.entries.get(&index) {
```

Comment:
Use .remove here instead of .get, no need to remove at the end.

Reply:
done